### ISSUE TITLE
Detections with runLen == 2 being displayed on public output

### ISSUE THREAD

**Issue by jsayers-bsc** [MotusDev/Motus-TO-DO#707]
Many of the false positives false detections described ([here ](https://github.com/MotusDev/Motus-TO-DO/issues/674)and [here](https://github.com/MotusDev/Motus-TO-DO/issues/610#issuecomment-590540710)) are being displayed in the public output. 

Glancing over a subset of these "detections" shows that many are runLen ==2, which should be filtered out of the public output. As discussed briefly on our call last week, somehow the filter that should be at work here just isn't catching these.

Since our call I've seen this several other times, including two others today ([here ](https://motus.org/data/receiverDeployment?id=3051)and [here](https://motus.org/data/receiverDeployment?id=5339)). Whatever the issue, it seems to be affecting a lot of stations, so I wanted to break this particular issue off from the main discussion in that hopes that it can be addressed.

**Comment by Hussell** [comment]
When logged-out (i.e. viewing it as a member of the public instead of an administrator) I see far fewer detections. Were you viewing these while logged-in?

**Comment by jsayers-bsc** [comment]
I didn't notice, though this sort of permission thing has been an thorn in the side previously!

I just checked now and I see the same number of detections logged in/out for this [one](https://motus.org/data/receiverDeployment?id=3051), and somewhat fewer detections while logged out for this [one ](https://motus.org/data/receiverDeployment?id=5339)(1247 tags vs 1328 tags, both crazy high numbers).

However, I don't think the permission issue should apply here though. Doesn't the run length filter on the detections table apply whether or not the user is logged in?

**Comment by Hussell** [comment]
Upon further investigation, I found three lines commented out in prTaghitsSummariesBatchID, the procedure that summarizes the data. The commented-out lines are for filtering out run-lengths <= 3, freqSD >= 0.1, and impossible timestamps. Unfortunately, there's no comment there about why they were commented out. So I'm going to punt this question to @cjardine-bsc, since she's likely the one who did it and thus knows the reasoning. (I suspect filtering through det_runs_filters is supposed to replace it, but hasn't? I haven't found anything there yet which replaces the commented-out filters.)

**Comment by cjardine-bsc** [comment]
Yes. Those are commented out so all the data can be in the runs summary for movebank. It should be filtered from web outputs using the det_runs_filters.

**Comment by denislepage** [comment]
Can you be more explicit about what is needed for this? Is there a view that already filters the data appropriately, or does the Java code need to be modified?

Thanks

**Comment by cjardine-bsc** [comment]
Sure thing. 

The view vwBase_taghits_runs should remove all filtered runs. It has a clause it in that only selects records where motusFilter is null. 

Any time we're pulling from det_taghits_runs directly we need to include this clause:

WHERE motusFilter is null


I just noticed there are a relatively small number of runs (300k) in det_taghits_runs that have len= 3 that do not have their motus filter populated. I'm tracing that back now.

**Comment by cjardine-bsc** [comment]
Oh they are CTT batches. All good.

**Comment by Hussell** [comment]
None of the runs this issue is about are from CTT batches:

select a.sensor_deploy_id, b.source, count(*) n
from det_taghits_runs a
join det_batches b on b.batchID = a.batch_id
where a.sensor_deploy_id in (3051, 5339)
and a.n_hits <= 3
and a.motusFilter is null
group by a.sensor_deploy_id, b.source

**Comment by cjardine-bsc** [comment]
Heya, 

On Friday, I finished re-running the summaries for the ~400 SG batches that were effected by this. Which means that the query above now returns zero records, which is good! Hopefully that has removed the runlength<=3 records for the sensors in this issue. 

However, over the weekend, 25 more batches were imported that did not have their summaries run properly. And now trigger this same issue (for different sensors): 

select *, dbo.ts2datetime(tsMotus) from det_batches 
where batchId in (988043,988030,987619,987610,987578,987566,987562,987539,987532,987500,987476,987448,
987444,987443,987436,987433,987423,987382,987371,987370,987369,987365,987364,987363,
987315) 

Which means this is a recurring problem. We had assumed on the call last week that the problem was from summaries occasionally failing due to locks or timeouts. But none of the summaries run this weekend failed. So there is something else happening. 

Ah. Got it. 

My clue was that this seemed to happen when jobs in the procs queue were ignored because they are duplicating jobs that are already running. 

SELECT *
  FROM [motus].[dbo].[procs_queue_steps]
  where name in ('prTaghitsSummariesAll','prSGBatchesActivityAll')  
  and job_id between 482003 and 483519 
  order by job_id desc

Remember that for the current system to work, the prSGBatchesActivityAll proc needs to run before the prTaghitsSummariesAll proc for a given batch ID. The first identifies bad runs, the second includes those flags in the summary. 

So I think the sequence of events that leads to this problem is as follows:

1. Batches are imported, initiating a job that runs both procs prSGBatchesActivityAll and prTaghitsSummariesAll (in proper order) via the procs queue. 

The SGBatchesActivityAll proc starts by pulling all new batches into a temp table to loop through:

insert into #temp (batchID)
SELECT batchID
FROM det_batches a WITH(NOLOCK) 
INNER JOIN lk_batch_filter z on a.batchFilter = z.batchFilter 
where batchID >= @start 
and z.exclude_batch = 0 
and tsMotus IS NOT NULL
and numhits > 0 
and source = 'sg'

2. BEFORE that proc is done running and looping through those batches, more batches are imported into the batches table. 

3. Then the prTaghitsSummariesAll proc starts. It also selects a list of batches into a temp table, but now there are more batches than were available when prSGBatchesActivityAll started. So this proc now runs a bunch of batches that have not yet been run through prSGBatchesActivityAll.

insert into #temp (batchID)
SELECT batchID
FROM det_batches a WITH (NOLOCK)
INNER JOIN lk_batch_filter z on a.batchFilter = z.batchFilter 
where a.numHits > 0 and batchID >= @start and z.exclude_batch = 0 
and tsMotus IS NOT NULL
and (@newonly = 0 OR is_new = 1)
and exists (select runID from det_runs where batchIDbegin = a.batchID EXCEPT select runID from det_runs_filters where batchIDbegin = a.batchID)
order by batchID

4. The next time prSGBatchesActivityAll is called, it runs some of the batches that have already been done by prTaghitsSummariesAll. 

Through this series of events we get cases where prTaghitsSummariesAll is run before prSGBatchesActivityAll. Which is how those bad runs are making it into the summary table. 

EUREKA!!!

Perhaps the simplest solution is to change prTaghitsSummariesAll so that it calls prSGBatchesActivityBatchID before making the summaries. That way the procs will always be run on the same temp table list of batches and we ensure they are always run sequentially. We could then remove prSGBatchesActivityAll from the list of procs run after import of SG data. 

@denislepage your blessing please?

**Comment by denislepage** [comment]
This may be happening, but in theory, it should not:

>BEFORE that proc is done running and looping through those batches, more batches are imported into the batches table

the 4 steps associated with post-processing a batch of imports (containing multiple batchID’s) are part of the same job_id. Both that job and the import job have a priority 3 (lowest), so only one of those should be running at the same time in the database. I can’t find evidence that those level 3 jobs ever overlap.

It’s a bit of a slow query, so I only limited this to the last 2 weeks, but this (I think) confirms that there are no overlapping priority 3 jobs that have run.

select * from procs_queue_jobs a inner join procs_queue_jobs b
on a.priority = b.priority and a.job_id < b.job_id
and (a.start_time <= b.end_time and a.end_time >= b.start_time)
where a.priority = 3 and a.start_time > '2020-03-15'
order by a.job_id desc

there may well be another flaw in the design, of course, but I don’t think that’s it.

As a safeguard, I think the methods that rely on the runs data for summaries should always explicitly exclude at least runs of len = 2 in addition to those in the filters.

**Comment by denislepage** [comment]
Have you already resolved the 25 batches that had issues? If yes, can you hold fixing more? What is the query you use to diagnose the problem?  Have you checked in sandbox if the same problem also exist, or whether the missed batches are the same?

**Comment by cjardine-bsc** [comment]
I left the 24 of the 25 batches (one I fixed while testing) because I assumed you'd want to look at this with things still broken. 

This is the query that will give you the batches. 

use motus
go 

select distinct  b.*
from det_taghits_runs a
join det_batches b on b.batchID = a.batch_id
where  a.n_hits <= 3
and a.motusFilter is null
and source = 'sg' 

I then compared the tsMotus from those batches to the begin and end time in the  proc_queue_steps table and realized the bad batches fell within instances when the duplicate proc was ignored. 

SELECT *
FROM [motus].[dbo].[procs_queue_steps]
where name in ('prTaghitsSummariesAll','prSGBatchesActivityAll')
and job_id between 482003 and 483519
order by job_id desc

I had not checked in sandbox, but I can. 

I'm running your query now, but if what you say is true and new batches cannot be imported until the batch summaries are done running, how do we get duplicate (ignored) entries into the queue for 'prTaghitsSummariesAll','prSGBatchesActivityAll'? Is there another process that is not priority 3 that would insert these into the procs queue?

Ie. Job ID 483409 

Select * from [motus].[dbo].[procs_queue_steps]
where job_id = 483409


RE: "As a safeguard, I think the methods that rely on the runs data for summaries should always explicitly exclude at least runs of len = 2 in addition to those in the filters."

Thats fine, but I don't think this is as good a solution as just ensuring that prSGBatchesActivity is run before summaries by calling it from the same proc (it runs very quickly, even for large batches). Otherwise you'll still end up with cases where runs that shouldn't be included due to hourbin noise make it through.

**Comment by denislepage** [comment]
Thats fine, but I don't think this is as good a solution as just ensuring that prSGBatchesActivity is run before summaries by calling it from the same proc (it runs very quickly, even for large batches). Otherwise you'll still end up with cases where runs that shouldn't be included due to hourbin noise make it through.
We can move to that, but my earlier point was that it should already not be happening, so we’ll want to make sure that they are not being missed for other reasons (e.g. the proc is called but ignored).

I’d want to try to replay the order in which things were suppose to happen and see if all adds up.

I can’t think why there are jobs that are ignored yet. Part of the issue is that we are relying on the @start parameter which is determined based on the highest value in the table, for efficiency, rather than getting an exhaustive list of batches that need to be run. I’d be perhaps inclined to change that behaviour, so missed batches all eventually get run, but we’ll have to do some testing.

Combining the 2 procs will probably also make sense, but may not entirely fix the issue.

**Comment by cjardine-bsc** [comment]
One thing that I should have pointed out. The summaries and batch activity for the bad batches were both populated. So they were run through both procs, just out of order.

**Comment by jsayers-bsc** [comment]
One of the stations mentioned earlier has another batch of these false positives showing up in the public detection tables: [https://motus.org/data/receiverDeploymentDetections?id=4462](https://motus.org/data/receiverDeploymentDetections?id=4462)

**Comment by cjardine-bsc** [comment]
yeah. 

I put in a solution to sandbox on Friday. I've been re-running all the bad batches in sandbox since,  (there were a couple thousand *wince* so it's still not completely re-run yet). Once that's done we'll be able to monitor sandbox for a couple of days and ensure that the fix I applied worked and there were not unintended consequences. After that I'll implement it in live. 

In the meantime, 92 new bad-batches have creeped into live over the past week or so. (fun fact, we passed batch 1,000,000 this week!). I've started re-running those 92 batches so those bad runs will be vanquished.
